diff -up scalapack-2.0.2/CMAKE/FortranMangling.cmake.orig scalapack-2.0.2/CMAKE/FortranMangling.cmake
--- scalapack-2.0.2/CMAKE/FortranMangling.cmake.orig	2012-01-14 00:41:00.000000000 +0300
+++ scalapack-2.0.2/CMAKE/FortranMangling.cmake	2019-08-18 23:39:31.444021933 +0300
@@ -24,7 +24,7 @@ FUNCTION(COMPILE RESULT)
     if(RESVAR EQUAL 0)
     MESSAGE(STATUS "Configure in the INSTALL directory successful")
     else()
-    MESSAGE(FATAL_ERROR " Configure in the BLACS INSTALL directory FAILED")
+    MESSAGE(STATUS " Configure in the BLACS INSTALL directory FAILED")
     MESSAGE(FATAL_ERROR " Output Build:\n ${LOG1}")
     endif()
 
@@ -36,7 +36,7 @@ FUNCTION(COMPILE RESULT)
     if(RESVAR  EQUAL 0)
     MESSAGE(STATUS "Build in the BLACS INSTALL directory successful")
     else()
-    MESSAGE(FATAL_ERROR " Build in the BLACS INSTALL directory FAILED")
+    MESSAGE(STATUS " Build in the BLACS INSTALL directory FAILED")
     MESSAGE(FATAL_ERROR " Output Build:\n ${LOG2}")
     endif()
     # Clean up:
